compiler/rustc_expand/src/base.rs

```diff
@@ -290,16 +290,15 @@ pub trait BangProcMacro {
 
 impl<F> BangProcMacro for F
 where
-    F: Fn(TokenStream) -> TokenStream,
+    F: Fn(&mut ExtCtxt<'_>, Span, TokenStream) -> TokenStream,
 {
     fn expand<'cx>(
         &self,
-        _ecx: &'cx mut ExtCtxt<'_>,
-        _span: Span,
+        ecx: &'cx mut ExtCtxt<'_>,
+        span: Span,
         ts: TokenStream,
     ) -> Result<TokenStream, ErrorGuaranteed> {
-        // FIXME setup implicit context in TLS before calling self.
-        Ok(self(ts))
+        Ok(self(ecx, span, ts))
```

Comment:
I think we can improve the Return types here to handle `ExpandResult::Retry(()) `. e.g.
```
pub enum BangRetry {
    Yes,
    No,
}

impl<F> BangProcMacro for F
where
    F: Fn(&mut ExtCtxt<'_>, Span, TokenStream) -> Result<TokenStream, BangRetry>,
{
    fn expand<'cx>(
        &self,
        _ecx: &'cx mut ExtCtxt<'_>,
        _span: Span,
        ts: TokenStream,
    ) -> Result<TokenStream, BangRetry> {
        // FIXME setup implicit context in TLS before calling self.
        self(ts)
    }
}
```

Reply:
It's a good suggestion. But I'll wait for feedback from @petrochenkov about whether converting `LegacyBang` to `Bang` is a good idea in general before putting more effort in.

@petrochenkov, what do you think?